odoo/orm/fields.py

```diff
@@ -1107,6 +1107,207 @@ def update_db_related(self, model):
+                "COALESCE(%(column)s->%(company_id)s,to_jsonb(%(fallback)s::%(column_type)s))",
+                column=sql_field,
+                company_id=str(model.env.company.id),
+                fallback=fallback,
+                column_type=SQL(self._column_type[1]),
+            )
+            if self.type in ('boolean', 'integer', 'float', 'monetary'):
+                return SQL('(%s)::%s', sql_field, SQL(self._column_type[1]))
+            # here the specified value for a company might be NULL e.g. '{"1": null}'::jsonb
+            # the result of current sql_field might be 'null'::jsonb
+            # ('null'::jsonb)::text == 'null'
+            # ('null'::jsonb->>0)::text IS NULL
+            return SQL('(%s->>0)::%s', sql_field, SQL(self._column_type[1]))
+        return sql_field
+
+    def property_to_sql(self, field_sql: SQL, property_name: str, model: BaseModel, alias: str, query: Query) -> SQL:
+        """ Return an :class:`SQL` object that represents the value of the given
+        expression from the given table alias.
+
+        The query object is necessary for fields that need to add tables to the query.
+        """
+        if not property_name:
+            return field_sql
+        raise ValueError(f"Invalid field property {property_name!r} on {self}")
+
+    def condition_to_sql(self, model: BaseModel, alias: str, field_expr: str, operator: str, value, query: Query) -> SQL:
+        """ Return an :class:`SQL` object that represents the domain condition
+        given by the triple ``(field_expr, operator, value)`` with the given
+        table alias, and in the context of the given query.
+
+        This method should use the model to resolve the SQL and check access
+        of the field.
+        """
+        sql_expr = self.__condition_to_sql(model, alias, field_expr, operator, value, query)
+        if self.company_dependent:
+            sql_expr = self._condition_to_sql_company(sql_expr, model, alias, field_expr, operator, value, query)
+        return sql_expr
+
+    def _condition_to_sql_company(self, sql_expr: SQL, model: BaseModel, alias: str, field_expr: str, operator: str, value, query: Query) -> SQL:
+        """ Add a not null condition on the field for company-dependent fields to use an existing index for better performance."""
+        if (
+            self.company_dependent
+            and self.index == 'btree_not_null'
+            and not (self.type in ('datetime', 'date') and field_expr != self.name)  # READ_GROUP_NUMBER_GRANULARITY is not supported
+            and model.env['ir.default']._evaluate_condition_with_fallback(model._name, field_expr, operator, value) is False
+        ):
+            return SQL('(%s IS NOT NULL AND %s)', SQL.identifier(alias, self.name), sql_expr)
+        return sql_expr
+
+    @classmethod
+    def _convert_sql_to_sql_text(cls, sql_field: SQL) -> SQL:
+        """ Cast the field to char."""
+        return SQL("%s::text", sql_field)
```

Comment:
There's only one call and one override that returns the input unchanged.
So this basically abstracts a boolean.
A simple `if self._is_sql_text` would have done the job, don't you think?